<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Кодировка-данных-и-создание-новой-выборки-с-признаками" data-toc-modified-id="Кодировка-данных-и-создание-новой-выборки-с-признаками-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Кодировка данных и создание новой выборки с признаками</a></span></li><li><span><a href="#Разделение-выборок-на-обучающие-и-тестовые" data-toc-modified-id="Разделение-выборок-на-обучающие-и-тестовые-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Разделение выборок на обучающие и тестовые</a></span></li><li><span><a href="#Проверка-качества-модели-линейной-регрессии-из-sklearn--до-и-после-преобразования-(метрика-R2)" data-toc-modified-id="Проверка-качества-модели-линейной-регрессии-из-sklearn--до-и-после-преобразования-(метрика-R2)-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Проверка качества модели линейной регрессии из sklearn  до и после преобразования (метрика R2)</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

## Загрузка данных

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
data = pd.read_csv('insurance.csv')

In [3]:
data.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
Пол                  5000 non-null int64
Возраст              5000 non-null float64
Зарплата             5000 non-null float64
Члены семьи          5000 non-null int64
Страховые выплаты    5000 non-null int64
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [5]:
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [6]:
data.duplicated().sum()

153

Зарплата и возраст - тип float. Тип данных заменю на int (так как отсутствуют дробные части).

Пропуски не выявлены. Числовые аномалии на первый взгляд отсутствуют. 

Есть 153 дубликата в данных (это около 3% данных) - можно и удалить, но выборка и так небольшая и задача поставлена на кодирование данных, а не на идеальную точность модели, хотя буду к этому стремиться.

Сразу выделю признаки - 'features', целевой признак (страховые выплаты) - 'target'.

In [7]:
data['Возраст'] = data['Возраст'].astype('int')
data['Зарплата'] = data['Зарплата'].astype('int')

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
Пол                  5000 non-null int64
Возраст              5000 non-null int64
Зарплата             5000 non-null int64
Члены семьи          5000 non-null int64
Страховые выплаты    5000 non-null int64
dtypes: int64(5)
memory usage: 195.4 KB


In [9]:
features = data.drop('Страховые выплаты', axis = 1)
target = data['Страховые выплаты']

In [10]:
features.shape, target.shape

((5000, 4), (5000,))

Предоставлены 5 000 данных о застрахованных (пол, возраст, количество членов семьи, зарплата, количнств страховых выплат)

Изменен тип данных на int, выделены признаки и целевой признак

## Умножение матриц

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
pred = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

Свойство обратной матрицы: $(AB)^{-1} = B^{-1}  A^{-1}$. Если матрица имеет обратную матрицу, значит она является обратимой

Свойство транспонированной матрицы: $(AB)^T = B^TA^T$

**Представлю новую матрицу $X1$ как произведение матрицы $X$ на обратимую матрицу $P$**

$$
X1 = XP
$$

Формула предсказания $pred1$ 

$$
pred1 = X1w1
$$

$$
pred1 = XPw1
$$

Рассчитаю $w1$ по формуле 

$$
w1 = (X1^TX1)^{-1}X1^Ty 
$$

$$
w1 = ((XP)^TXP)^{-1}(XP)^Ty 
$$

$$
w1 = (P^T (X^T X) P)^{-1} (XP)^T y
$$

Раскрою скобки $(P^T (X^T X) P)^{-1}$ и $(XP)^T$:

$$
w1 = P^{-1}(X^T X)^{-1}(P^T)^{-1}(P^TX^T)y
$$   


Матрица $P$ - квадратная и обратимая, соответственно $(P^T)^{-1} P^T = E$.

$E$ - единичная матрица, умножение на которую не меняят любую матрицу. Соответственно из формулы можно убрать $E$ и получится следующее:
    
$$
w1 = P^{-1}(X^T X)^{-1}EX^Ty = P^{-1}(X^T X)^{-1}X^Ty
$$   


В левой части можно вычленить часть формулы $(X^T X)^{-1}X^Ty$, которая является формулой для вычисления $w$ (вектора весов), соотвественно заменю эту часть на $w$:
    
$$
w1 = P^{-1}w
$$    

    
Подставлю полученнное значение $w1$ в формулу для получения предсказания:
    
$$
pred1 = XPP^{-1}w
$$

$PP^{-1}$ - это единичная матрица $E$, при умножении на которую матрица $X$ не изменится:
    
$$
pred1 = Xw
$$   

**Ответ:** Качество линейной регресии не изменится при умножении матрицы $X$ на обратимую матрицу $P$, предсказания не поменяются. 

**Обоснование:** В результате преобразовании формулы (см. выше) $pred1$ равно $pred$. 

Однако, вектора весов до преобразования ($w$) и после ($w1$) отличаются, в дальнейшем нужно будет еще раз обучить модель:

$$
w = (X^T X)^{-1} X^T y
$$

$$
w1 = P^{-1}w
$$

## Алгоритм преобразования

**Алгоритм**

Сначала проверю работу модели (построенной на основании библиотеки sklearn и на основании линейной алгебры) на данных перед преобразованием, рассчитаю метрику r2

Затем осуществлю преобразования данных клиентов, для чего:

1) создам случайную квадратную матрицу $P$ с помощью функции numpy.random.normal(), 

2) проверю случайную матрицу на обратимость, путем создания обратной матрицы $P^{-1}$ с помощью функции numpy.linalg.inv(),

3) создам новую матрицу features_encoded путем произведения матриц features и $P$

4) обучу модель на новых данных 

5) проверю метрику r2 для преобразованных данных

6) сравню метрики r2 для модели, обученной до преобразования данных и после преобразования

**Обоснование**

Ранее в разделе 3 я выяснила, что признаки (матрица $X$) умножаются на обратимую матрицу ($P$), предсказания модели не изменилось.

Соотвественно предложенный алгоритм не должен влиять на работу модели.

В качестве обоснования проведу расчет предсказаний на основании формул 


$$
w = (X^T X)^{-1} X^T y
$$

$$
pred = Xw
$$

Для сокращения формулы предсказания в матрицы добавлю столбец, состоящий только из единиц (он идёт нулевым)

In [11]:
X = features.values
X1 = np.concatenate((np.ones((features.shape[0], 1)), features), axis=1)

In [12]:
y = target
w = np.linalg.inv(X1.T.dot(X1)) @ X1.T @ y
pred = X1.dot(w)

Создаю случайную квадратную матрицу $P$ (4x4), в которой количество признаков равно количеству признаков в исходной матрице, и проверю ее на обратимость

In [13]:
P = np.random.normal(size = (4,4))
P

array([[-0.71596715, -0.07512468, -1.05197922,  1.43457217],
       [ 1.66054415, -0.69107283, -0.9398518 ,  1.4565793 ],
       [ 1.60001233,  0.22752007,  0.4752715 ,  0.17024542],
       [-0.88461679, -1.62807973,  0.13148543, -0.6568961 ]])

Найду обратную ей матрицу

In [14]:
P_inverted = np.linalg.inv(P)
P_inverted

array([[-0.545087  ,  0.47006809, -0.2201934 , -0.20514979],
       [-0.17301054, -0.07064904, -0.34971357, -0.62512041],
       [ 1.40088933, -1.2920071 ,  2.42238538,  0.8223006 ],
       [ 1.44324891, -0.71653335,  1.64813997,  0.46787514]])

Случайная матрица $P$ является обратимой.

Далее создам новую матрицу X_new путем произведения матриц $X$ и $P$, обучу модель по формулам на преобразованных данных и сравню векторы предсказаний $pred$ и $pred1$

In [15]:
X_new = X.dot(P)
X_new = np.concatenate((np.ones((X_new.shape[0], 1)), X_new), axis=1)
X_new

array([[1.00000000e+00, 7.94270934e+04, 1.12549584e+04, 2.35340118e+04,
        8.50467031e+03],
       [1.00000000e+00, 6.08759690e+04, 8.61234531e+03, 1.80172151e+04,
        6.53567175e+03],
       [1.00000000e+00, 3.36484148e+04, 4.75788040e+03, 9.95344571e+03,
        3.61739464e+03],
       ...,
       [1.00000000e+00, 5.42718597e+04, 7.69585283e+03, 1.60931696e+04,
        5.79913756e+03],
       [1.00000000e+00, 5.23535654e+04, 7.41974339e+03, 1.55200437e+04,
        5.59853389e+03],
       [1.00000000e+00, 6.50053954e+04, 9.21626169e+03, 1.92687864e+04,
        6.95352599e+03]])

In [16]:
y = target
w1 = np.linalg.inv(X_new.T.dot(X_new)) @ X_new.T @ y
pred1 = X_new.dot(w1)

In [17]:
pred

array([ 0.51172714,  0.68431581,  0.09373368, ..., -0.25697045,
       -0.1909916 ,  0.04904991])

In [18]:
pred1

array([ 0.51172722,  0.68431577,  0.09373364, ..., -0.25697046,
       -0.19099147,  0.04905   ])

Как видно, различия в предсказаниях есть, но они незначительные. Алгоритм преобразования можно считать удачным

## Проверка алгоритма ##

### Кодировка данных и создание новой выборки с признаками ###

In [19]:
features_encoded = pd.DataFrame(data = features.values @ P, columns = features.columns)
features_encoded.head()

Пол       Возраст      Зарплата  Члены семьи
0  79427.093417  11254.958388  23534.011778  8504.670306
1  60875.969048   8612.345311  18017.215146  6535.671747
2  33648.414762   4757.880403   9953.445711  3617.394639
3  66753.616458   9469.818319  19799.347461  7128.508426
4  41806.101147   5918.848719  12377.218213  4485.624279

Кодирование данных произвелось успешно

### Разделение выборок на обучающие и тестовые ###

In [20]:
train_features, test_features, train_target, test_target = train_test_split(features, 
                                                                            target, 
                                                                            test_size = 0.25, 
                                                                            random_state = 12345)

In [21]:
train_features.shape, test_features.shape, train_target.shape, test_target.shape

((3750, 4), (1250, 4), (3750,), (1250,))

In [22]:
train_features_encoded, test_features_encoded, train_target_encoded, test_target_encoded = train_test_split(features_encoded, 
                                                                            target, 
                                                                            test_size = 0.25, 
                                                                            random_state = 12345)

In [23]:
train_features_encoded.shape, test_features_encoded.shape, train_target_encoded.shape, test_target_encoded.shape

((3750, 4), (1250, 4), (3750,), (1250,))

Разделение произвелось удачно

### Проверка качества модели линейной регрессии из sklearn  до и после преобразования (метрика R2)  ###

In [24]:
model = LinearRegression()
model.fit(train_features, train_target)
predictions = model.predict(test_features)

In [25]:
print('R2 на исходных данных: ', r2_score(test_target, predictions))

R2 на исходных данных:  0.43522756840832477


In [26]:
model_encoded = LinearRegression()
model_encoded.fit(train_features_encoded, train_target_encoded)
predictions_encoded = model_encoded.predict(test_features_encoded)

In [27]:
print('R2 на преобразованных  данных: ', r2_score(test_target_encoded, predictions_encoded))

R2 на преобразованных  данных:  0.43522756840833987


Метрика R2 на преобразованных данных практически идентична метрике R2 на исходных данных.

Это говорит о том, что качество модели не изменилось. Модель рабочая, так как 0 < метрика R2 < 1, однако она ближе к 0, значит над параметрами еще можно поработать.

## Общий вывод ##

Во время работы над проектом были загружены и изучены данные о клиентах страховой компании.

Произведена замена типа данных у 2-х признаков.

Проверено умножается ли матрица признаков на обратимую матрицу, а также предложен и проверен алгоритм преобразования данных с целью их защиты.

Алгоритм работает, данные выводятся в измененном виде, слабо похожем на реальность.

Обучены 2 модели линейной регрессии (на исходных данных и на преобразованных данных). 

Обе модели показали практически идентичный показатель метрик R2 - 0.435. 

Показатель метрики выше 0 и меньше 1, что говорит о том, что модели рабочие. 

Однако показатель ближе к 0, значит работа моделей не идеальна, можно ее улучшить: лучше поработав с исходными данными (в них есть порядка 3 % полных дубликатов, их можно удалить или проанализировать и произвести корректные замены), изменив гиперпараметры